src/webcore/executor.rs

```diff
@@ -0,0 +1,226 @@
+    // queue of pending tasks.
+    microtask_queue: RefCell< VecDeque< Rc< SpawnedTask > > >,
+    waker: Reference,
+    shrink_counter: Cell<usize>
+}
+
+// Not strictly necessary, but may become relevant in the future
+impl Drop for EventLoopInner {
+    #[inline]
+    fn drop( &mut self ) {
+        js! { @(no_return)
+            @{&self.waker}.drop();
+        }
+    }
+}
+
+impl EventLoopInner {
+    // Initializes the event loop. Only called once.
+    fn new() -> Self {
+        EventLoopInner {
+            microtask_queue: RefCell::new(VecDeque::with_capacity(INITIAL_QUEUE_CAPACITY)),
+            waker: js!(
+                var callback = @{|| EventLoop.drain()};
+                var wrapper = function() {
+                    if (!callback.dropped) { callback() }
+                };
+
+                // Modern browsers can use `MutationObserver` which allows
+                // us to schedule a micro-task without allocating a promise.
+                // https://dom.spec.whatwg.org/#notify-mutation-observers
+                if ( typeof MutationObserver === "function" ) {
+                    var node = document.createTextNode( "0" );
+                    var state = false;
+
+                    new MutationObserver( wrapper ).observe( node, { characterData: true } );
+
+                    function nextTick() {
+                        state = !state;
+                        node.data = ( state ? "1" : "0" );
+                    }
+
+                // Node.js and other environments
+                } else {
+                    var promise = Promise.resolve( null );
+
+                    function nextTick() {
+                        promise.then( wrapper );
+                    }
+                }
+
+                nextTick.drop = function() {
+                    callback.dropped = true;
+                    callback.drop();
+                };
+
+                return nextTick;
+            ).try_into().unwrap(),
+            shrink_counter: Cell::new(0)
+        }
+    }
+    // Pushes a task onto the queue
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        queue.push_back(task);
+
+        // If the queue was previously empty, then we need to schedule
+        // the queue to be drained.
+        if queue.len() == 1 {
+            self.wake();
+        }
+    }
+    // Invoke the JavaScript waker function
+    fn wake(&self) {
+        js! { @(no_return) @{&self.waker}(); }
+    }
+    // Remove and return a task from the front of the queue
+    fn pop_task(&self) -> Option< Rc< SpawnedTask > > {
+        self.microtask_queue.borrow_mut().pop_front()
+    }
+    fn shrink_if_necessary(&self) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        // We consider shrinking the queue if it is less than
+        // half full...
+        if queue.len() <= queue.capacity() / 2 {
```

Comment:
This is wrong. The `len` will always be `0`, because it's called after draining. It should instead shrink it if the capacity is larger than `INITIAL_QUEUE_CAPACITY`